src/main/java/pl/zaradny/springApp/domain/Product.java

```diff
@@ -56,57 +68,108 @@ public boolean equals(Object o) {
                 Objects.equals(price, product.price) &&
                 Objects.equals(image, product.image) &&
                 Objects.equals(description, product.description) &&
-                Objects.equals(createdAt, product.createdAt);
+                Objects.equals(createdAt, product.createdAt) &&
+                Objects.equals(tags, product.tags);
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(id, name, price, image, description, createdAt);
+        return Objects.hash(id, name, price, image, description, createdAt, tags);
     }
 
-    @Override
-    public String toString() {
-        return "Product{" +
-                "id='" + id + '\'' +
-                ", name='" + name + '\'' +
-                ", price=" + price +
-                ", image=" + image +
-                ", description=" + description +
-                ", createdAt=" + createdAt +
-                '}';
+    public static ProductBuilder build(){
+        return new ProductBuilder();
     }
 
-    public static final class ProductBuilder{
-
-        //required
-        private final String id;
-        private final String name;
-        private final Price price;
-        private final LocalDateTime createdAt;
+    public static class ProductBuilder{
 
-        //optional
+        private String id;
+        private String name;
+        private Price price;
+        private LocalDateTime createdAt;
         private Image image;
         private Description description;
+        private Set<Tag> tags;
+
+        private ProductBuilder() {}
 
-        public ProductBuilder(String id, String name, Price price, LocalDateTime createdAt) {
+        public ProductBuilder withTags(Set<Tag> tags){
+            this.tags = tags;
+            return this;
+        }
+
+        public ProductBuilder withId(String id) {
             this.id = id;
+            return this;
+        }
+
+        public ProductBuilder withName(String name) {
             this.name = name;
+            return this;
+        }
+
+        public ProductBuilder withPrice(Price price) {
             this.price = price;
+            return this;
+        }
+
+        public ProductBuilder withCreatedAt(LocalDateTime createdAt) {
             this.createdAt = createdAt;
+            return this;
         }
 
-        public ProductBuilder setImage(Image newImage){
+        public ProductBuilder withImage(Image newImage){
             this.image = newImage;
             return this;
         }
 
-        public ProductBuilder setDescription(Description description){
+        public ProductBuilder withDescription(Description description){
             this.description = description;
             return this;
         }
 
+        private boolean isNameValid() {
+            try{
```

Comment:
ponownie jak przy cenie, jeżeli rozważysz całą tą konstrukcję to może się okazać, że proste sprawdzenie tego null-a będzie ładniejsze :D 

```
if(this.name == null || this.name.isEmpty()) {
    throw new EmptyProductNameException();
}
return true;
```

Tylko popatrzmy dalej, masz metodę która zaczyna się od `is` zwraca boolean a jednak nie ma szansy zwrócić false :D

dla takiej metody dajesz aplikacji szansę sprawdzić czy jest poprawny bez rzucenia wyjątkiem i rozwalenia flow, jeżeli bardzo chcesz rzucać wyjątkiem to zmień nazwę na `validateName` która jest void wtedy dla potencjalnej osoby korzystającej z twojej domeny będzie to dużo bardziej przewidywalne co się stanie